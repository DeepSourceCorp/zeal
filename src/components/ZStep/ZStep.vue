<template>
  <div
    class="mr-0 z-step"
    :class="{
      'w-1/3 flex-shrink-0 flex-grow-0': align === 'horizontal' && isLast,
      'w-1/2': align == 'horizontal' && !isLast,
      'flex flex-row': align === 'vertical'
    }"
  >
    <div
      class="relative z-step__head"
      :class="{
        currentStatus: true,
        'w-full': align === 'horizontal',
        'flex flex-col justify-center': align === 'vertical' && !isLast
      }"
    >
      <div
        v-if="align == 'horizontal' && !isLast"
        class="z-step__line absolute h-0.5 top-2.5 rounded-sm w-auto left-7 right-1"
        :class="{
          'bg-juniper': currentStatus === 'completed',
          'bg-ink-200': currentStatus !== 'completed'
        }"
      ></div>
      <div
        class="w-6 h-6 rounded-full z-step__icon"
        :class="{
          'bg-juniper': currentStatus == 'completed',
          'bg-robin': currentStatus == 'active',
          'bg-ink-200': !showNumbers && currentStatus == 'default',
          'border border-vanilla-200': showNumbers && currentStatus == 'default',
          'flex-shrink-0': align === 'vertical'
        }"
      >
        <span v-if="showNumbers" class="flex items-center justify-center h-full text-base font-bold text-vanilla-100">{{
          index + 1
        }}</span>
        <span
          v-else-if="currentStatus == 'completed'"
          class="flex items-center justify-center h-full text-base font-bold text-ink-400"
          >âœ“</span
        >
      </div>
      <div
        v-if="align == 'vertical' && !isLast"
        class="z-step__line h-full w-0.5 my-1 rounded-sm mx-auto"
        :class="{
          'bg-juniper': currentStatus === 'completed',
          'bg-ink-200': currentStatus !== 'completed'
        }"
      ></div>
    </div>
    <div class="z-step__main" :class="{ 'mt-4': align === 'horizontal', 'mb-4 ml-4 w-full': align === 'vertical' }">
      <slot name="title">
        <div class="mt-1 text-xs font-medium leading-snug tracking-widest uppercase z-step__title text-vanilla-400">
          {{ title }}
        </div>
      </slot>
      <div
        class="mt-1 text-sm font-bold leading-6 z-step__description"
        :class="{
          'text-vanilla-200': currentStatus,
          'text-vanilla-400': currentStatus == 'default'
        }"
      >
        <slot name="description" v-bind:status="currentStatus">{{ description }}</slot>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'ZStep',
  props: {
    title: {
      required: false,
      default: undefined,
      type: String
    },
    description: {
      required: false,
      default: undefined,
      type: String
    },
    status: {
      required: false,
      type: String,
      default: 'default',
      validator: value => {
        return ['default', 'active', 'completed'].includes(value)
      }
    }
  },
  data() {
    return {
      index: -1,
      showNumbers: false,
      internalStatus: 'default',
      align: 'horizontal'
    }
  },
  computed: {
    currentStatus() {
      return this.status || this.internalStatus
    },
    isLast() {
      const parent = this.$parent
      return parent.steps[parent.steps.length - 1] === this
    }
  },
  beforeCreate() {
    const parent = this.$parent
    parent.steps.push(this)
  },
  beforeDestroy() {
    const parent = this.$parent,
      steps = parent.steps
    const index = steps.indexOf(this)
    if (index >= 0) {
      steps.slice(index, 1)
    }
  }
}
</script>
