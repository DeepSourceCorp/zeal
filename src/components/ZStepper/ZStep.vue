<template>
  <div class="z-step w-1/2 mr-0" :class="[isLast && 'w-1/3 flex-shrink-0 flex-grow-0']">
    <div class="z-step__head relative w-full" :class="[`${currentStatus}`]">
      <div
        class="z-step__line absolute h-0.5 top-2.5 rounded-sm w-auto left-7 right-1"
        :class="[(currentStatus == 'completed' && 'bg-juniper') || 'bg-ink-200']"
        v-if="!isLast"
      ></div>
      <div class="z-step__icon rounded-full w-6 h-6" :class="[getIconStyle]">
        <span
          v-if="currentStatus == 'completed'"
          class="text-vanilla-100 flex justify-center items-center text-base h-full font-bold"
          >âœ“</span
        >
      </div>
    </div>
    <div class="z-step__main">
      <div
        class="z-step__title uppercase text-slate font-medium mt-4 tracking-wider leading-snug text-sm"
      >
        <slot name="title">{{ title }}</slot>
      </div>
      <div class="z-step__description font-bold leading-6 mt-1" :class="[getTextStyle]">
        <slot name="description">{{ description }}</slot>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'

const bgColors = {
  completed: 'bg-juniper',
  active: 'bg-robin',
  default: 'bg-ink-200'
}

export default Vue.extend({
  name: 'ZStep',
  props: {
    title: {
      type: String
    },
    description: {
      type: String
    },
    status: {
      type: String
    }
  },
  data() {
    return {
      index: -1,
      internalStatus: ''
    }
  },
  computed: {
    currentStatus(): any {
      return this.status || this.internalStatus
    },
    getIconStyle(): string {
      return bgColors[this.currentStatus] || bgColors['default']
    },
    getTextStyle(): string {
      const activeColor = 'text-vanilla-200',
        defaultColor = 'text-vanilla-400'
      return (this.currentStatus && activeColor) || defaultColor
    },
    isLast(): boolean {
      const parent: any = this.$parent
      return parent.steps[parent.steps.length - 1] === this
    }
  },
  beforeCreate() {
    const parent: any = this.$parent
    parent.steps.push(this)
  },
  beforeDestroy() {
    const parent: any = this.$parent,
      steps = parent.steps
    const index = steps.indexOf(this)
    if (index >= 0) {
      steps.slice(index, 1)
    }
  }
})
</script>
