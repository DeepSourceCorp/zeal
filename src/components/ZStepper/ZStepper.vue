<template>
  <div class="z-stepper flex">
    <slot></slot>
  </div>
</template>

<script lang="ts">
export default {
  name: 'ZStepper',
  props: {
    active: {
      default: 0,
      type: Number
    }
  },
  data() {
    return {
      steps: [],
      stepOffSet: 0
    }
  },
  watch: {
    steps(steps: any) {
      steps.forEach((step: any, index: number) => {
        step.index = index
        if (index === this.active) {
          step.internalStatus = 'active'
        } else if (index < this.active) {
          step.internalStatus = 'completed'
        }
      })
    }
  }
}
</script>
