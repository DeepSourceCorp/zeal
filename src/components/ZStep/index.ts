import ZStep from '../ZStepper/ZStep.vue'

export default ZStep
