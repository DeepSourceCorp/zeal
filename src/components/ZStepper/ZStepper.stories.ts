import '../../assets/css/tailwind.css'
import '../../assets/css/typography.css'
import '../../assets/css/layout.css'
import ZStepper from './ZStepper.vue'
import ZStep from './ZStep.vue'
import ZButton from '../ZButton/ZButton.vue';

export default {
  title: 'Stepper',
  component: ZStepper,
  excludeStories: /.*Data$/
}

export const DefaultStepper = () => ({
  components: { ZStepper, ZStep },
    template: `<div class="container">
        <z-stepper>
            <z-step title="Step 1" description="Some description"></z-step>
            <z-step title="Step 2" description="Some description"></z-step>
            <z-step title="Step 3" description="Some description"></z-step>
        </z-stepper>
    </div>`
})

export const StepperWithDirectStatus = () => ({
  components: { ZStepper, ZStep },
    template: `<div class="container">
        <z-stepper>
            <z-step title="Step 1" description="Some description" status="completed"></z-step>
            <z-step title="Step 2" description="Some description" status="active"></z-step>
            <z-step title="Step 3" description="Some description"></z-step>
        </z-stepper>
    </div>`
})

export const StepperWithStatusFromParent = () => ({
  components: { ZStepper, ZStep, ZButton },
  template: `<div class="container">
      <z-stepper :active="2">
          <z-step title="Step 1" description="Some description"></z-step>
          <z-step title="Step 2" description="Some description"></z-step>
          <z-step title="Step 3" description="Some description"></z-step>
      </z-stepper>
  </div>`
})
