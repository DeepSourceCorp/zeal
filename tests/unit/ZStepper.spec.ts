import ZStepper from '../../src/components/ZStepper'
import ZStep from '../../src/components/ZStep'
import Vue from 'vue'

import { createLocalVue, mount, Wrapper } from '@vue/test-utils'

let id = 0;

const createElm = function() {
  const elm = document.createElement('div');

  elm.id = 'app' + ++id;
  document.body.appendChild(elm);

  return elm;
};

export const createVue = function(Compo: any) {
  return new Vue({ template: Compo }).$mount(createElm());
};

const div = document.createElement('div')
div.id = 'root'
document.body.appendChild(div)

const Stepper = {
    template: `<el-steps>
        <el-step title="step1"></el-step>
        <el-step title="step2"></el-step>
        <el-step title="step3"></el-step>
      </el-steps>`
}

describe('ZStepper', () => {
  let mountFn: (options?: object) => Wrapper<Vue>
  let localVue: typeof Vue

  beforeEach(() => {
    mountFn = (options = {}) => {
      return mount(Stepper, {
        localVue,
        ...options
      })
    }
  })

  it('renders a default stepper component', () => {
      let wrapper = mountFn({
        attachTo: '#root'
    })
      expect(wrapper.vm.$el.querySelectorAll('.z-step')).toHaveLength(3);
      wrapper.destroy()
  })
})
