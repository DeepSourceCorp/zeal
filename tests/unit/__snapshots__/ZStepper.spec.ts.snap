// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZStepper renders a default stepper component 1`] = `
<div class="z-stepper flex">
  <div class="mr-0 z-step w-1/3 flex-shrink-0 flex-grow-0">
    <div class="relative z-step__head currentStatus w-full">
      <!---->
      <div class="w-6 h-6 rounded-full z-step__icon bg-ink-200">
        <!---->
      </div>
      <!---->
    </div>
    <div class="z-step__main mt-4">
      <div class="mt-1 text-xs font-medium leading-snug tracking-widest uppercase z-step__title text-vanilla-400">
        step1
      </div>
      <div class="mt-1 text-sm font-bold leading-6 z-step__description text-vanilla-200 text-vanilla-400"></div>
    </div>
  </div>
  <div class="mr-0 z-step w-1/3 flex-shrink-0 flex-grow-0">
    <div class="relative z-step__head currentStatus w-full">
      <!---->
      <div class="w-6 h-6 rounded-full z-step__icon bg-ink-200">
        <!---->
      </div>
      <!---->
    </div>
    <div class="z-step__main mt-4">
      <div class="mt-1 text-xs font-medium leading-snug tracking-widest uppercase z-step__title text-vanilla-400">
        step2
      </div>
      <div class="mt-1 text-sm font-bold leading-6 z-step__description text-vanilla-200 text-vanilla-400"></div>
    </div>
  </div>
  <div class="mr-0 z-step w-1/3 flex-shrink-0 flex-grow-0">
    <div class="relative z-step__head currentStatus w-full">
      <!---->
      <div class="w-6 h-6 rounded-full z-step__icon bg-ink-200">
        <!---->
      </div>
      <!---->
    </div>
    <div class="z-step__main mt-4">
      <div class="mt-1 text-xs font-medium leading-snug tracking-widest uppercase z-step__title text-vanilla-400">
        step3
      </div>
      <div class="mt-1 text-sm font-bold leading-6 z-step__description text-vanilla-200 text-vanilla-400"></div>
    </div>
  </div>
</div>
`;

exports[`ZStepper renders a stepper component with status from the Step 1`] = `
<div class="z-stepper flex">
  <div class="mr-0 z-step w-1/3 flex-shrink-0 flex-grow-0">
    <div class="relative z-step__head currentStatus w-full">
      <!---->
      <div class="w-6 h-6 rounded-full z-step__icon bg-juniper"><span class="flex items-center justify-center h-full"><svg viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" fill="none" class="stroke-2 z-icon w-4 h-4 z-icon--check text-ink-400"><polyline points="20 6 9 17 4 12"></polyline></svg></span></div>
      <!---->
    </div>
    <div class="z-step__main mt-4">
      <div class="mt-1 text-xs font-medium leading-snug tracking-widest uppercase z-step__title text-vanilla-400">
        Step 1
      </div>
      <div class="mt-1 text-sm font-bold leading-6 z-step__description text-vanilla-200">Some description</div>
    </div>
  </div>
  <div class="mr-0 z-step w-1/3 flex-shrink-0 flex-grow-0">
    <div class="relative z-step__head currentStatus w-full">
      <!---->
      <div class="w-6 h-6 rounded-full z-step__icon bg-robin"><span class="flex items-center justify-center h-full"><svg viewBox="0 0 24 24" stroke-linecap="round" stroke-linejoin="round" fill="none" class="stroke-2 z-icon w-4 h-4 z-icon--circle-dashed text-vanilla-100"><path d="M12 22C17.5228 22 22 17.5228 22 12C22 6.47715 17.5228 2 12 2C6.47715 2 2 6.47715 2 12C2 17.5228 6.47715 22 12 22Z" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" stroke-dasharray="4 4"></path></svg></span></div>
      <!---->
    </div>
    <div class="z-step__main mt-4">
      <div class="mt-1 text-xs font-medium leading-snug tracking-widest uppercase z-step__title text-vanilla-400">
        Step 2
      </div>
      <div class="mt-1 text-sm font-bold leading-6 z-step__description text-vanilla-200">Some description</div>
    </div>
  </div>
  <div class="mr-0 z-step w-1/3 flex-shrink-0 flex-grow-0">
    <div class="relative z-step__head currentStatus w-full">
      <!---->
      <div class="w-6 h-6 rounded-full z-step__icon bg-ink-200">
        <!---->
      </div>
      <!---->
    </div>
    <div class="z-step__main mt-4">
      <div class="mt-1 text-xs font-medium leading-snug tracking-widest uppercase z-step__title text-vanilla-400">
        Step 3
      </div>
      <div class="mt-1 text-sm font-bold leading-6 z-step__description text-vanilla-200 text-vanilla-400">Some description</div>
    </div>
  </div>
</div>
`;

exports[`ZStepper renders a stepper component with status passed from the Parent Component 1`] = `
<div class="z-stepper flex">
  <div class="mr-0 z-step w-1/3 flex-shrink-0 flex-grow-0">
    <div class="relative z-step__head currentStatus w-full">
      <!---->
      <div class="w-6 h-6 rounded-full z-step__icon bg-ink-200">
        <!---->
      </div>
      <!---->
    </div>
    <div class="z-step__main mt-4">
      <div class="mt-1 text-xs font-medium leading-snug tracking-widest uppercase z-step__title text-vanilla-400">
        Step 1
      </div>
      <div class="mt-1 text-sm font-bold leading-6 z-step__description text-vanilla-200 text-vanilla-400">Some description</div>
    </div>
  </div>
  <div class="mr-0 z-step w-1/3 flex-shrink-0 flex-grow-0">
    <div class="relative z-step__head currentStatus w-full">
      <!---->
      <div class="w-6 h-6 rounded-full z-step__icon bg-ink-200">
        <!---->
      </div>
      <!---->
    </div>
    <div class="z-step__main mt-4">
      <div class="mt-1 text-xs font-medium leading-snug tracking-widest uppercase z-step__title text-vanilla-400">
        Step 2
      </div>
      <div class="mt-1 text-sm font-bold leading-6 z-step__description text-vanilla-200 text-vanilla-400">Some description</div>
    </div>
  </div>
  <div class="mr-0 z-step w-1/3 flex-shrink-0 flex-grow-0">
    <div class="relative z-step__head currentStatus w-full">
      <!---->
      <div class="w-6 h-6 rounded-full z-step__icon bg-ink-200">
        <!---->
      </div>
      <!---->
    </div>
    <div class="z-step__main mt-4">
      <div class="mt-1 text-xs font-medium leading-snug tracking-widest uppercase z-step__title text-vanilla-400">
        Step 3
      </div>
      <div class="mt-1 text-sm font-bold leading-6 z-step__description text-vanilla-200 text-vanilla-400">Some description</div>
    </div>
  </div>
</div>
`;

exports[`ZStepper renders a vertical stepper component 1`] = `
<div class="z-stepper flex flex-col">
  <div class="mr-0 z-step w-1/3 flex-shrink-0 flex-grow-0">
    <div class="relative z-step__head currentStatus w-full">
      <!---->
      <div class="w-6 h-6 rounded-full z-step__icon bg-ink-200">
        <!---->
      </div>
      <!---->
    </div>
    <div class="z-step__main mt-4">
      <div class="mt-1 text-xs font-medium leading-snug tracking-widest uppercase z-step__title text-vanilla-400">
        Step 1
      </div>
      <div class="mt-1 text-sm font-bold leading-6 z-step__description text-vanilla-200 text-vanilla-400">Some description</div>
    </div>
  </div>
  <div class="mr-0 z-step w-1/3 flex-shrink-0 flex-grow-0">
    <div class="relative z-step__head currentStatus w-full">
      <!---->
      <div class="w-6 h-6 rounded-full z-step__icon bg-ink-200">
        <!---->
      </div>
      <!---->
    </div>
    <div class="z-step__main mt-4">
      <div class="mt-1 text-xs font-medium leading-snug tracking-widest uppercase z-step__title text-vanilla-400">
        Step 2
      </div>
      <div class="mt-1 text-sm font-bold leading-6 z-step__description text-vanilla-200 text-vanilla-400">Some description</div>
    </div>
  </div>
  <div class="mr-0 z-step w-1/3 flex-shrink-0 flex-grow-0">
    <div class="relative z-step__head currentStatus w-full">
      <!---->
      <div class="w-6 h-6 rounded-full z-step__icon bg-ink-200">
        <!---->
      </div>
      <!---->
    </div>
    <div class="z-step__main mt-4">
      <div class="mt-1 text-xs font-medium leading-snug tracking-widest uppercase z-step__title text-vanilla-400">
        Step 3
      </div>
      <div class="mt-1 text-sm font-bold leading-6 z-step__description text-vanilla-200 text-vanilla-400">Some description</div>
    </div>
  </div>
</div>
`;
